File: packages/forma-typescript/src/parser.ts
import type { FormaProgram, FormaTask } from "./types.js";

const blockNames = ["input", "output", "compute", "agent", "permissions", "constraints", "verify"] as const;

export function parseForma(source: string): FormaProgram {
  const tasks = parseTasks(source);
  if (tasks.length === 0) {
    throw new Error("F0001: expected task declaration");
  }

  return { tasks };
}

function parseTasks(source: string): FormaTask[] {
  const tasks: FormaTask[] = [];
  const taskStart = /task\s+([A-Za-z_][A-Za-z0-9_]*)\s*\{/g;
  let match: RegExpExecArray | null;
  while ((match = taskStart.exec(source))) {
    const name = match[1];
    const openBrace = taskStart.lastIndex - 1;
    const closeBrace = findMatchingBrace(source, openBrace);
    if (closeBrace === -1) {
      throw new Error("F0001: expected task declaration");
    }
    const body = source.slice(openBrace + 1, closeBrace);
    tasks.push(parseTask(name, body, {
      start: positionAt(source, match.index),
      end: positionAt(source, closeBrace + 1),
    }));
    taskStart.lastIndex = closeBrace + 1;
  }
  return tasks;
}

function parseTask(name: string | undefined, body: string, sourceSpan: NonNullable<FormaTask["sourceSpan"]>): FormaTask {
  if (!name || !body) {
    throw new Error("F0001: expected task declaration");
  }

  const intent = extractIntent(body);
  const input = parseFieldBlock(extractBlock(body, "input"));
  const output = parseFieldBlock(extractBlock(body, "output"));
  const task: FormaTask = {
    name,
    intent,
    input: input.fields,
    output: output.fields,
    schemas: output.schemas,
    compute: lines(extractBlock(body, "compute", false)),
    permissions: lines(extractBlock(body, "permissions", false)),
    constraints: lines(extractBlock(body, "constraints", false)),
    verify: lines(extractBlock(body, "verify", false)),
    sourceSpan,
  };

  const agent = extractBlock(body, "agent", false);
  if (agent.length > 0) {
    task.agentInstruction = extractInstruction(agent);
  }

  return task;
}

function positionAt(source: string, index: number): { line: number; column: number } {
  const before = source.slice(0, index);
  const line = before.split("\n").length;
  const lastNewline = before.lastIndexOf("\n");
  return { line, column: index - lastNewline };
}

function findMatchingBrace(source: string, openBrace: number): number {
  let depth = 0;
  for (let index = openBrace; index < source.length; index += 1) {
    const char = source[index];
    if (char === "{") depth += 1;
    if (char === "}") {
      depth -= 1;
      if (depth === 0) return index;
    }
  }
  return -1;
}

function extractIntent(body: string): string {
  const match = body.match(/intent\s+"([^"]*)"/);
  const intent = match?.[1];
  if (!intent) throw new Error("F1001: task requires intent");
  return intent;
}

function extractBlock(body: string, name: (typeof blockNames)[number], required = true): string {
  const match = new RegExp(`\\b${name}\\s*\\{`, "m").exec(body);
  if (!match) {
    if (required) throw new Error(`F1002: task requires ${name} block`);
    return "";
  }
  const openBrace = match.index + match[0].lastIndexOf("{");
  const closeBrace = findMatchingBrace(body, openBrace);
  if (closeBrace === -1) {
    throw new Error(`F1002: task requires ${name} block`);
  }
  return body.slice(openBrace + 1, closeBrace);
}

function parseFieldBlock(block: string): { fields: FormaTask["input"]; schemas: FormaTask["schemas"] } {
  const schemas: FormaTask["schemas"] = {};
  let fieldSource = block;
  const objectStart = /\bobject\s+([A-Za-z_][A-Za-z0-9_]*)\s*\{/g;
  let match: RegExpExecArray | null;
  const ranges: Array<[number, number]> = [];
  while ((match = objectStart.exec(block))) {
    const schemaName = match[1];
    if (!schemaName) {
      throw new Error(`F1003: invalid field declaration '${match[0].trim()}'`);
    }
    const openBrace = objectStart.lastIndex - 1;
    const closeBrace = findMatchingBrace(block, openBrace);
    if (closeBrace === -1) {
      throw new Error(`F1003: invalid field declaration '${match[0].trim()}'`);
    }
    schemas[schemaName] = parseFields(block.slice(openBrace + 1, closeBrace));
    ranges.push([match.index, closeBrace + 1]);
    objectStart.lastIndex = closeBrace + 1;
  }
  for (const [start, end] of ranges.reverse()) {
    fieldSource = `${fieldSource.slice(0, start)}${fieldSource.slice(end)}`;
  }
  return { fields: parseFields(fieldSource), schemas };
}

function parseFields(block: string): FormaTask["input"] {
  const fields: FormaTask["input"] = {};
  for (const line of lines(block)) {
    const match = line.match(/^([A-Za-z_][A-Za-z0-9_]*)\s*:\s*([A-Za-z_][A-Za-z0-9_]*)(\[\])?(\?)?$/);
    const name = match?.[1];
    const type = match?.[2];
    if (!name || !type) throw new Error(`F1003: invalid field declaration '${line}'`);
    fields[name] = { type, array: Boolean(match?.[3]), optional: Boolean(match?.[4]) };
  }
  return fields;
}

function extractInstruction(block: string): string {
  const match = block.match(/instruction\s+"""([\s\S]*?)"""/);
  const instruction = match?.[1];
  if (instruction === undefined) throw new Error("F1004: agent block requires instruction");
  return instruction.trim();
}

function lines(block: string): string[] {
  return block
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean);
}

File: packages/forma-python/src/forma/parser.py
import re

from .types import FormaProgram, FormaTask


def parse_forma(source: str) -> FormaProgram:
    tasks = _tasks(source)
    if not tasks:
        raise ValueError("F0001: expected task declaration")
    return FormaProgram(tasks=tasks)


def _tasks(source: str) -> list[FormaTask]:
    tasks: list[FormaTask] = []
    for match in re.finditer(r"task\s+([A-Za-z_][A-Za-z0-9_]*)\s*\{", source):
        open_brace = match.end() - 1
        close_brace = _matching_brace(source, open_brace)
        if close_brace == -1:
            raise ValueError("F0001: expected task declaration")
        body = source[open_brace + 1 : close_brace]
        tasks.append(_task(match.group(1), body))
    return tasks


def _task(name: str, body: str) -> FormaTask:
    agent = _block(body, "agent", required=False)
    parsed_input = _field_block(_block(body, "input"))
    parsed_output = _field_block(_block(body, "output"))
    return FormaTask(
        name=name,
        intent=_intent(body),
        input=parsed_input["fields"],
        output=parsed_output["fields"],
        schemas=parsed_output["schemas"],
        compute=_lines(_block(body, "compute", required=False)),
        permissions=_lines(_block(body, "permissions", required=False)),
        constraints=_lines(_block(body, "constraints", required=False)),
        verify=_lines(_block(body, "verify", required=False)),
        agent_instruction=_instruction(agent) if agent else None,
    )


def _matching_brace(source: str, open_brace: int) -> int:
    depth = 0
    for index in range(open_brace, len(source)):
        char = source[index]
        if char == "{":
            depth += 1
        if char == "}":
            depth -= 1
            if depth == 0:
                return index
    return -1


def _intent(body: str) -> str:
    match = re.search(r'intent\s+"([^"]*)"', body)
    if not match:
        raise ValueError("F1001: task requires intent")
    return match.group(1)


def _block(body: str, name: str, required: bool = True) -> str:
    match = re.search(rf"\b{name}\s*\{{", body)
    if not match:
        if required:
            raise ValueError(f"F1002: task requires {name} block")
        return ""
    open_brace = match.end() - 1
    close_brace = _matching_brace(body, open_brace)
    if close_brace == -1:
        raise ValueError(f"F1002: task requires {name} block")
    return body[open_brace + 1 : close_brace]


def _field_block(block: str) -> dict[str, dict[str, object]]:
    schemas: dict[str, dict[str, object]] = {}
    field_source = block
    ranges: list[tuple[int, int]] = []
    for match in re.finditer(r"\bobject\s+([A-Za-z_][A-Za-z0-9_]*)\s*\{", block):
        open_brace = match.end() - 1
        close_brace = _matching_brace(block, open_brace)
        if close_brace == -1:
            raise ValueError(f"F1003: invalid field declaration '{match.group(0).strip()}'")
        schemas[match.group(1)] = _fields(block[open_brace + 1 : close_brace])
        ranges.append((match.start(), close_brace + 1))
    for start, end in reversed(ranges):
        field_source = f"{field_source[:start]}{field_source[end:]}"
    return {"fields": _fields(field_source), "schemas": schemas}


def _fields(block: str) -> dict[str, dict[str, object]]:
    fields: dict[str, dict[str, object]] = {}
    for line in _lines(block):
        match = re.match(r"^([A-Za-z_][A-Za-z0-9_]*)\s*:\s*([A-Za-z_][A-Za-z0-9_]*)(\[\])?(\?)?$", line)
        if not match:
            raise ValueError(f"F1003: invalid field declaration '{line}'")
        fields[match.group(1)] = {"type": match.group(2), "array": bool(match.group(3)), "optional": bool(match.group(4))}
    return fields


def _instruction(block: str) -> str:
    match = re.search(r'instruction\s+"""([\s\S]*?)"""', block)
    if not match:
        raise ValueError("F1004: agent block requires instruction")
    return match.group(1).strip()


def _lines(block: str) -> list[str]:
    return [line.strip() for line in block.splitlines() if line.strip()]
